rockets/jsonrpc/receiverInterface.h

```diff
@@ -0,0 +1,251 @@
+     * @throw std::invalid_argument if the method name starts with "rpc."
+     */
+    void bindAsync(const std::string& method, DelayedResponseCallback action);
+
+    /**
+     * Bind a method to an async response callback with templated parameters.
+     *
+     * The parameters object must be deserializable by a free function:
+     * from_json(Params& object, const std::string& json).
+     *
+     * @param method to register.
+     * @param action to perform that will notify the caller upon completion.
+     * @throw std::invalid_argument if the method name starts with "rpc."
+     */
+    template <typename Params>
+    void bindAsync(const std::string& method,
+                   std::function<void(Params, AsyncResponse)> action)
+    {
+        bindAsync(method,
+                  [action](const Request& request, AsyncResponse callback) {
+                      Params params;
+                      if (from_json(params, request.message))
+                          action(std::move(params), callback);
+                      else
+                          callback(Response::invalidParams());
+                  });
+    }
+
+    /**
+     * Process a JSON-RPC request and block for the result.
+     *
+     * @param request Request object with message in JSON-RPC 2.0 format.
+     * @return json response string in JSON-RPC 2.0 format.
+     */
+    std::string process(const Request& request);
+
+    /**
+     * Process a JSON-RPC request asynchronously.
+     *
+     * @param request Request object with message in JSON-RPC 2.0 format.
+     * @return future json response string in JSON-RPC 2.0 format.
+     */
+    std::future<std::string> processAsync(const Request& request);
+
+    /**
+     * Process a JSON-RPC request asynchronously.
+     *
+     * @param request Request object with message in JSON-RPC 2.0 format.
+     * @param callback that return a json response string in JSON-RPC 2.0
+     *        format upon request completion.
+     */
+    void process(const Request& request, AsyncStringResponse callback);
+
+    /** @internal */
+    bool hasMethod(const std::string& method) const;
+    const DelayedResponseCallback& getMethod(const std::string& method) const;
+
+protected:
+    /**
+     * Process a JSON-RPC request asynchronously.
+     *
+     * @param request Request object with message in JSON-RPC 2.0 format.
+     * @param callback that return a json response string in JSON-RPC 2.0
+     *        format upon request completion.
+     * @sa RequestProcessor::process
+     */
+    virtual void _process(const Request& request,
+                          AsyncStringResponse callback) = 0;
+
+    /**
+     * Callback for each method added by bind() or connect() to reject the name
+     * by throwing a std::invalid_argument exception
+     */
+    virtual void _filterMethodRegistration(const std::string&) {}
+private:
+    std::map<std::string, DelayedResponseCallback> _methods;
```

Comment:
In order to make this class purely abstract base class, I'm wondering if we could move the _methods map to implementation classes, replacing _filterMethodRegistration() by a virtual register(). What do you think?